File: backend/utils/apify_all_jobs.py
"""
Server-side client for Apify actor agentx/all-jobs-scraper (All Jobs Scraper).
Used by scheduled backfill; keep token in APIFY_TOKEN (never VITE_*).
"""

from __future__ import annotations

import os
import re
from typing import Any
from urllib.parse import urlencode

import httpx

from models.schemas import JobListing

APIFY_ALL_JOBS_ACTOR_ID = "jpraRc4MCUh5ehbHV"
RUN_SYNC = f"https://api.apify.com/v2/acts/{APIFY_ALL_JOBS_ACTOR_ID}/run-sync-get-dataset-items"

APIFY_ALLOWED_COUNTRIES = frozenset(
    {
        "Argentina",
        "Australia",
        "Austria",
        "Bahrain",
        "Bangladesh",
        "Belgium",
        "Bulgaria",
        "Brazil",
        "Canada",
        "Chile",
        "China",
        "Colombia",
        "Costa Rica",
        "Croatia",
        "Cyprus",
        "Czech Republic",
        "Denmark",
        "Ecuador",
        "Egypt",
        "Estonia",
        "Finland",
        "France",
        "Germany",
        "Greece",
        "Hong Kong",
        "Hungary",
        "India",
        "Indonesia",
        "Ireland",
        "Israel",
        "Italy",
        "Japan",
        "Kuwait",
        "Latvia",
        "Lithuania",
        "Luxembourg",
        "Malaysia",
        "Malta",
        "Mexico",
        "Morocco",
        "Netherlands",
        "New Zealand",
        "Nigeria",
        "Norway",
        "Oman",
        "Pakistan",
        "Panama",
        "Peru",
        "Philippines",
        "Poland",
        "Portugal",
        "Qatar",
        "Romania",
        "Saudi Arabia",
        "Singapore",
        "Slovakia",
        "Slovenia",
        "South Africa",
        "South Korea",
        "Spain",
        "Sweden",
        "Switzerland",
        "Taiwan",
        "Thailand",
        "Turkey",
        "Ukraine",
        "United Arab Emirates",
        "United Kingdom",
        "United States",
        "Uruguay",
        "Venezuela",
        "Vietnam",
    }
)

_COUNTRY_ALIASES = {
    "usa": "United States",
    "us": "United States",
    "u.s.a": "United States",
    "u.s.a.": "United States",
    "u.s": "United States",
    "u.s.": "United States",
    "america": "United States",
    "uk": "United Kingdom",
    "u.k": "United Kingdom",
    "u.k.": "United Kingdom",
    "britain": "United Kingdom",
    "england": "United Kingdom",
    "uae": "United Arab Emirates",
    "korea": "South Korea",
}


def _strip_quotes(s: str) -> str:
    t = s.strip()
    while (t.startswith('"') and t.endswith('"')) or (t.startswith("'") and t.endswith("'")):
        t = t[1:-1].strip()
    return t


def resolve_country_for_apify(raw: str | None, fallback: str = "United States") -> str:
    cleaned = _strip_quotes((raw or "").strip())
    if not cleaned:
        return fallback
    lower = cleaned.lower()
    if lower in _COUNTRY_ALIASES:
        return _COUNTRY_ALIASES[lower]
    if cleaned in APIFY_ALLOWED_COUNTRIES:
        return cleaned
    canon = next((c for c in APIFY_ALLOWED_COUNTRIES if c.lower() == lower), None)
    if canon:
        return canon
    print(f"[apify] Unknown country {cleaned!r} — using {fallback!r}. Fix JOBS_BACKFILL_COUNTRY.")
    return fallback


def _skills_list(skills: Any) -> list[str]:
    if isinstance(skills, list):
        return [str(s).strip() for s in skills if s]
    if isinstance(skills, str) and skills.strip():
        return [p.strip() for p in re.split(r"[,;]", skills) if p.strip()]
    return []


def _salary_range(item: dict[str, Any]) -> str | None:
    lo, hi = item.get("salary_minimum"), item.get("salary_maximum")
    cur = item.get("salary_currency") or ""
    period = item.get("salary_period") or ""
    if isinstance(lo, (int, float)) and isinstance(hi, (int, float)):
        return f"{lo}-{hi} {cur} ({period})".strip()
    return None


def apify_record_to_job_listing(item: dict[str, Any]) -> JobListing:
    platform_url = str(item.get("platform_url") or "")
    official_url = str(item.get("official_url") or "")
    title = str(item.get("title") or "")
    company = str(item.get("company_name") or "")
    posted = str(item.get("posted_date") or "")
    jid = platform_url or official_url or f"apify:{item.get('platform','job')}:{title}:{company}:{posted}"

    return JobListing(
        id=jid[:2048],
        job_title=title,
        company_name=company,
        location=str(item.get("location") or "Remote"),
        job_description=str(item.get("description") or ""),
        salary_range=_salary_range(item),
        job_listing_link=(official_url or platform_url or "#")[:4096],
        remote_allowed=bool(item.get("is_remote")),
        experience_level=str(item.get("experience_range") or "") or None,
        skills_required=_skills_list(item.get("skills")),
    )


def fetch_apify_jobs(
    *,
    keyword: str,
    country: str,
    token: str,
    max_results: int = 50,
    remote_only: bool = False,
    job_type: str = "all",
    posted_since: str = "7 days",
    location: str | None = None,
    max_total_charge_usd: float = 5.0,
    timeout_sec: int = 300,
) -> list[JobListing]:
    """
    Run All Jobs Scraper synchronously and return JobListing rows.
    """
    token = (token or "").strip()
    if not token:
        raise RuntimeError("APIFY_TOKEN is not set")

    country = resolve_country_for_apify(country)
    body: dict[str, Any] = {
        "keyword": keyword,
        "country": country,
        "max_results": min(1000, max(1, max_results)),
        "remote_only": remote_only,
        "posted_since": posted_since,
        "job_type": job_type,
        "currency": "USD",
    }
    if location and location.strip():
        body["location"] = location.strip()

    params = {
        "timeout": str(timeout_sec),
        "limit": str(max_results),
        "offset": "0",
        "maxTotalChargeUsd": str(max_total_charge_usd),
    }
    url = f"{RUN_SYNC}?{urlencode(params)}"

    with httpx.Client(timeout=timeout_sec + 30) as client:
        r = client.post(
            url,
            json=body,
            headers={
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
            },
        )
    if r.status_code >= 400:
        detail = r.text[:2000]
        try:
            err = r.json()
            if isinstance(err, dict) and err.get("error"):
                detail = str(err["error"].get("message") or err["error"])
        except Exception:
            pass
        raise RuntimeError(f"Apify HTTP {r.status_code}: {detail}")

    data = r.json()
    rows: list[dict[str, Any]]
    if isinstance(data, list):
        rows = data
    elif isinstance(data, dict) and isinstance(data.get("items"), list):
        rows = data["items"]
    else:
        rows = []

    return [apify_record_to_job_listing(row) for row in rows if isinstance(row, dict)]


def get_apify_token() -> str:
    return (os.getenv("APIFY_TOKEN") or os.getenv("APIFY_API_TOKEN") or "").strip()

File: frontend/src/lib/apifyAllJobs.ts
import axios from "axios";
import type { Job } from "../types/job";

/** [All Jobs Scraper](https://apify.com/agentx/all-jobs-scraper) — actor id from Apify Console. */
export const APIFY_ALL_JOBS_ACTOR_ID = "jpraRc4MCUh5ehbHV";

const RUN_SYNC = `https://api.apify.com/v2/acts/${APIFY_ALL_JOBS_ACTOR_ID}/run-sync-get-dataset-items`;

/** Matches prior live-search cap; keeps each Apify run small (pay-per-event per job). */
export const ALL_JOBS_MAX_RESULTS_PER_SEARCH = 10;

/** Actor `country` must match this list exactly (see Apify input schema). */
export const APIFY_ALLOWED_COUNTRIES = [
  "Argentina",
  "Australia",
  "Austria",
  "Bahrain",
  "Bangladesh",
  "Belgium",
  "Bulgaria",
  "Brazil",
  "Canada",
  "Chile",
  "China",
  "Colombia",
  "Costa Rica",
  "Croatia",
  "Cyprus",
  "Czech Republic",
  "Denmark",
  "Ecuador",
  "Egypt",
  "Estonia",
  "Finland",
  "France",
  "Germany",
  "Greece",
  "Hong Kong",
  "Hungary",
  "India",
  "Indonesia",
  "Ireland",
  "Israel",
  "Italy",
  "Japan",
  "Kuwait",
  "Latvia",
  "Lithuania",
  "Luxembourg",
  "Malaysia",
  "Malta",
  "Mexico",
  "Morocco",
  "Netherlands",
  "New Zealand",
  "Nigeria",
  "Norway",
  "Oman",
  "Pakistan",
  "Panama",
  "Peru",
  "Philippines",
  "Poland",
  "Portugal",
  "Qatar",
  "Romania",
  "Saudi Arabia",
  "Singapore",
  "Slovakia",
  "Slovenia",
  "South Africa",
  "South Korea",
  "Spain",
  "Sweden",
  "Switzerland",
  "Taiwan",
  "Thailand",
  "Turkey",
  "Ukraine",
  "United Arab Emirates",
  "United Kingdom",
  "United States",
  "Uruguay",
  "Venezuela",
  "Vietnam",
] as const;

export type ApifyCountry = (typeof APIFY_ALLOWED_COUNTRIES)[number];

const ALLOWED_COUNTRY_BY_LOWER = new Map(
  APIFY_ALLOWED_COUNTRIES.map((c) => [c.toLowerCase(), c] as const)
);

/** Common `.env` / user typos → exact Apify label */
const COUNTRY_ALIASES: Record<string, ApifyCountry> = {
  usa: "United States",
  us: "United States",
  "u.s.a": "United States",
  "u.s.a.": "United States",
  "u.s": "United States",
  "u.s.": "United States",
  america: "United States",
  uk: "United Kingdom",
  "u.k": "United Kingdom",
  "u.k.": "United Kingdom",
  britain: "United Kingdom",
  england: "United Kingdom",
  scotland: "United Kingdom",
  wales: "United Kingdom",
  "great britain": "United Kingdom",
  uae: "United Arab Emirates",
  emirates: "United Arab Emirates",
  korea: "South Korea",
  "south-korea": "South Korea",
};

function stripSurroundingQuotes(s: string): string {
  let t = s.trim();
  while (
    (t.startsWith('"') && t.endsWith('"')) ||
    (t.startsWith("'") && t.endsWith("'"))
  ) {
    t = t.slice(1, -1).trim();
  }
  return t;
}

/**
 * Resolves `VITE_JOBS_SEARCH_COUNTRY` or UI input to a value Apify accepts.
 * Unknown values fall back so a bad `.env` does not break every search.
 */
export function resolveCountryForApify(
  raw: string | undefined | null,
  fallback: ApifyCountry = "United States"
): ApifyCountry {
  const cleaned = stripSurroundingQuotes(String(raw ?? "").trim());
  if (!cleaned) return fallback;
  const lower = cleaned.toLowerCase();
  const alias = COUNTRY_ALIASES[lower];
  if (alias) return alias;
  const exact = ALLOWED_COUNTRY_BY_LOWER.get(lower);
  if (exact) return exact;
  if (import.meta.env.DEV && cleaned) {
    console.warn(
      `[Apify All Jobs Scraper] Unknown country "${cleaned}" — using "${fallback}". ` +
        `Use an exact label from the actor schema (e.g. United States, United Kingdom) or remove VITE_JOBS_SEARCH_COUNTRY.`
    );
  }
  return fallback;
}

function bearer(): string {
  const token =
    import.meta.env.VITE_APIFY_TOKEN?.trim() ||
    import.meta.env.VITE_APIFY_API_TOKEN?.trim();
  if (!token) {
    throw new Error(
      "Missing VITE_APIFY_TOKEN in .env — create a free account at https://console.apify.com/ → Integrations → API token."
    );
  }
  return token;
}

function formatAxiosError(err: unknown): string {
  if (!axios.isAxiosError(err)) return String(err ?? "Unknown error");
  const status = err.response?.status;
  const statusText = err.response?.statusText;
  const payload = err.response?.data as unknown;
  let detail = "";
  if (payload && typeof payload === "object" && "error" in payload) {
    const e = (payload as { error?: { message?: string; type?: string } }).error;
    if (e?.message) detail = e.message;
    else if (e?.type) detail = e.type;
  } else {
    try {
      if (typeof payload === "string") detail = payload;
      else if (payload && typeof payload === "object") detail = JSON.stringify(payload);
    } catch {
      /* ignore */
    }
  }
  const head = status
    ? `Apify All Jobs Scraper failed (${status}${statusText ? ` ${statusText}` : ""})`
    : "Apify All Jobs Scraper request failed";
  const bits = [head, err.message, detail].map((s) => (s || "").trim()).filter(Boolean);
  return bits.join(" — ");
}

/** Map dashboard/Jobs employment filter to actor `job_type` (schema: all | fulltime | contract). */
export function mapEmploymentTypeToApifyJobType(employmentType: string): string {
  const u = employmentType.trim().toUpperCase();
  if (u === "FULLTIME") return "fulltime";
  if (u === "CONTRACTOR") return "contract";
  /** Part-time / intern filters are not separate enums on this actor — widen search then filter in UI. */
  return "all";
}

function skillsFromApifyField(skills: unknown): string[] | undefined {
  if (Array.isArray(skills)) return skills.map(String).filter(Boolean);
  if (typeof skills === "string" && skills.trim()) {
    return skills
      .split(/[,;]/)
      .map((s) => s.trim())
      .filter(Boolean);
  }
  return undefined;
}

export function normalizeApifyAllJobsItem(item: Record<string, unknown>): Job {
  const platformUrl = item.platform_url != null ? String(item.platform_url) : "";
  const officialUrl = item.official_url != null ? String(item.official_url) : "";
  const title = String(item.title ?? "");
  const company = String(item.company_name ?? "");
  const posted = item.posted_date != null ? String(item.posted_date) : undefined;
  const id =
    platformUrl ||
    officialUrl ||
    `apify:${String(item.platform ?? "job")}:${title}:${company}:${posted ?? ""}`;

  return {
    id,
    title,
    company,
    employerLogo: item.company_logo != null ? String(item.company_logo) : undefined,
    location: String(item.location ?? ""),
    description: String(item.description ?? ""),
    applyUrl: officialUrl || platformUrl || undefined,
    salaryMin: typeof item.salary_minimum === "number" ? item.salary_minimum : undefined,
    salaryMax: typeof item.salary_maximum === "number" ? item.salary_maximum : undefined,
    remote: !!item.is_remote,
    skills: skillsFromApifyField(item.skills),
    postedAt: posted,
  };
}

export type SearchApifyAllJobsParams = {
  query?: string;
  page?: number;
  remoteOnly?: boolean;
  employmentType?: string;
  /** City/region (no country — use `country`). */
  location?: string;
  /** e.g. `United States` — default from env or United States. */
  country?: string;
};

/**
 * Runs [All Jobs Scraper](https://apify.com/agentx/all-jobs-scraper) synchronously and returns normalized jobs.
 * Works on Apify’s **Free** plan (sign up at https://apify.com/pricing — monthly usage credits).
 * The actor bills pay-per-event per job scraped; we cap `max_results` and `maxTotalChargeUsd` per request.
 */
export async function searchApifyAllJobs(
  params: SearchApifyAllJobsParams = {}
): Promise<Job[]> {
  const {
    query = "Software Engineer",
    page = 1,
    remoteOnly = false,
    employmentType = "",
    location = "",
    country: countryParam,
  } = params;

  const country = resolveCountryForApify(
    countryParam || import.meta.env.VITE_JOBS_SEARCH_COUNTRY,
    "United States"
  );

  const jobType = mapEmploymentTypeToApifyJobType(employmentType);

  const maxResults = ALL_JOBS_MAX_RESULTS_PER_SEARCH;
  const offset = Math.max(0, (page - 1) * ALL_JOBS_MAX_RESULTS_PER_SEARCH);

  const input: Record<string, unknown> = {
    keyword: query,
    country,
    max_results: Math.min(250, maxResults + offset + 5),
    remote_only: remoteOnly,
    posted_since: "1 month",
    job_type: jobType,
    currency: "USD",
  };
  if (location.trim()) input.location = location.trim();

  const qs = new URLSearchParams({
    timeout: "300",
    limit: String(maxResults),
    offset: String(offset),
    /** Hard cap on actor charges for one request (Apify may require ≥ $1 on some pricing tiers). */
    maxTotalChargeUsd: "1",
  });

  try {
    const { data } = await axios.post<Record<string, unknown>[] | { items?: Record<string, unknown>[] }>(
      `${RUN_SYNC}?${qs}`,
      input,
      {
        headers: {
          Authorization: `Bearer ${bearer()}`,
          "Content-Type": "application/json",
        },
        timeout: 320_000,
      }
    );
    const raw = Array.isArray(data) ? data : Array.isArray(data?.items) ? data.items : [];
    return raw.map((row) => normalizeApifyAllJobsItem(row)).slice(0, ALL_JOBS_MAX_RESULTS_PER_SEARCH);
  } catch (err: unknown) {
    throw new Error(formatAxiosError(err));
  }
}
